p2panda-rs/src/document/document_view.rs

```diff
@@ -4,71 +4,87 @@
+
+use crate::document::DocumentId;
+use crate::hash::{Hash, HashError};
 use crate::operation::OperationValue;
+use crate::Validate;
 
-/// The ID of a document view. Contains the hash id of the document, and the hash ids of the
-/// current document graph tips.
-#[derive(Debug, PartialEq, Clone)]
-pub struct DocumentViewId {
-    document_id: Hash,
-    view_id: Vec<Hash>,
-}
+/// The identifier of a document view.
+///
+/// Contains the hashes of the document graph tips which is all the information we need to reliably
+/// recreate the document at this certain point in time.
+#[derive(Debug, Eq, PartialEq, Clone, Serialize, Deserialize)]
+pub struct DocumentViewId(Vec<Hash>);
 
 impl DocumentViewId {
     /// Create a new document view id.
-    pub fn new(document_id: Hash, view_id: Vec<Hash>) -> Self {
-        Self {
-            document_id,
-            view_id,
-        }
+    pub fn new(graph_tips: Vec<Hash>) -> Self {
+        Self(graph_tips)
     }
 
-    /// Get just the document id.
-    pub fn document_id(&self) -> &Hash {
-        &self.document_id
+    /// Get the graph tip hashes of this view id.
+    pub fn graph_tips(&self) -> &[Hash] {
+        self.0.as_slice()
     }
+}
+
+impl Validate for DocumentViewId {
+    type Error = HashError;
+
+    fn validate(&self) -> Result<(), Self::Error> {
+        for hash in &self.0 {
+            hash.validate()?;
+        }
 
-    /// Get just the view id.
-    pub fn view_id(&self) -> &[Hash] {
-        self.view_id.as_slice()
+        Ok(())
     }
 }
 
 type FieldKey = String;
 
 /// The materialised view of a `Document`. It's fields match the documents schema definition.
 ///
-/// `DocumentViews` can be instantiated from a CREATE operation and then mutated with UPDATE
-/// or DELETE operations.
+/// `DocumentViews` are immutable versions of a `Document`. They represent a document at a certain
+/// point in time.
 #[derive(Debug, PartialEq, Clone)]
 pub struct DocumentView {
+    /// Identifier of this document view.
     pub(crate) id: DocumentViewId,
+
+    /// Identifier of the document this view is derived from.
+    pub(crate) document_id: DocumentId,
+
+    /// Materialized data held by this document view.
     pub(crate) view: BTreeMap<FieldKey, OperationValue>,
 }
 
 impl DocumentView {
     /// Construct a document view.
     ///
-    /// Requires the DocumentViewId and field values to be calculated seperately and then passed in
-    /// during construction.
-    pub fn new(id: DocumentViewId, view: BTreeMap<FieldKey, OperationValue>) -> Self {
-        Self { id, view }
+    /// Requires the DocumentId, DocumentViewId and field values to be calculated seperately and
+    /// then passed in during construction.
+    pub fn new(
+        id: DocumentViewId,
+        document_id: DocumentId,
+        view: BTreeMap<FieldKey, OperationValue>,
+    ) -> Self {
+        Self {
+            id,
+            document_id,
+            view,
+        }
     }
```

Comment:
Does this seem like an ok constructor pattern to you? Not related to the changes in this PR really. The fact it's possible to pass in any combination of `DocumentViewId`, `DocumentId` and `view` seems quite haphazard. Should we make sure these structs are all private to the document construction process (where the validation occurs).

Reply:
Making something private / wrapping everything around something which protects from misuse can definitely help here

Reply:
Let's make a ticket for this?